### PR TITLE
Dynamic update of Kubespray and K8s [#10]

### PR DESCRIPTION
**Pull request by michaelspedersen**
## Description
Adds dynamic update of Kubespray at provisioning stage.
The version will be the most recent release of Kubespray. In addition, the Kubernetes version is based on the most recent supported version in Kubespray.

The Kubernetes version is fetched during the `generate_config` stage, while Kubespray is updated in the `provision` stage.

To utilize the dynamic update, release type must be specified as "kubespray" when generating the configuration file
```
/k8s-infra/bin/k8sinfra generate_config (...) --release-type=kubespray (...)
```

In addition to the dynamic update, a few patches to the Kubespray code are also included.

## Context
With the existing image, the newest (stable) Kubernetes release is deployed. While this might be unofficially supported in the most recent Kubespray release, this cannot be guaranteed.

The changes are done based on observations in [CNF Testbed](https://github.com/cncf/cnf-testbed), which deploys Kubernetes with Multus and CNI plugins by default. Due to this, a small set of patches are included to the Kubespray code, when deploying with the "kubespray" release-type.

## Issues:
https://github.com/crosscloudci/k8s-infra/issues/10
https://github.com/cncf/cnf-testbed/issues/344
https://github.com/cncf/cnf-testbed/issues/345

## How Has This Been Tested?
* [ ]  Covered by existing integration testing
* [ ]  Added integration testing to cover
* [ ] Tested with [trigger client](https://github.com/crosscloudci/crosscloudci-trigger) against 
   * [ ] cidev.cncf.ci
   * [ ] dev.cncf.ci
   * [ ] staging.cncf.ci
   * [ ] cncf.ci (production)
* [x]  Tested locally
* [ ]  Have not tested

## Types of changes
* [x]  Bug fix (non-breaking change which fixes an issue)
* [x]  New feature (non-breaking change which adds functionality)
* [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
 
## Checklist:
* [x]  My change requires a change to the documentation.
* [ ]  I have updated the documentation accordingly.
* [ ] No updates required.

### CODE REVIEW

File: bin/k8sinfra [hunk 1]
```diff
@@ -153,6 +153,10 @@ class K8sInfra < Thor
       if options['dry-run'] then
         puts "Dry-run successfully completed"
       else
+        if @cluster_hash['k8s_infra']["release_type"]=="kubespray" then
+	  latest = ks.latest_kubespray_release
```
Comment (taylor) [inline bin/k8sinfra:157]: indent these lines

File: bin/k8sinfra [hunk 2]
```diff
@@ -153,6 +153,10 @@ class K8sInfra < Thor
...
+	  ks.update_kubespray(latest)
```
Comment (taylor) [inline bin/k8sinfra:158]: updating kubespray itself every time might be a good idea even if using the latest stable K8s release from upstream.

File: bin/k8sinfra [hunk 3]
```diff
@@ -233,8 +237,17 @@ class K8sInfra < Thor
...
-    kubernetes_release = K8sUtils.kubernetes_release("#{options['release-type']}/#{options['arch']}")
-    stable_k8s_release = K8sUtils.kubernetes_release('stable')
+    unless options['release-type'] == "kubespray"
+      kubernetes_release = K8sUtils.kubernetes_release("#{options['release-type']}/#{options['arch']}")
+      stable_k8s_release = K8sUtils.kubernetes_release('stable')
+    else
+      ks = Kubespray.new(@cluster_hash)
+      ks_version = ks.latest_kubespray_release
+      k8s_version = ks.latest_supported_kubernetes(ks_version)
+      kubernetes_release = k8s_version + "\n"
```
Comment (taylor) [inline bin/k8sinfra:247]: are the newlines necessary?

File: bin/k8sinfra (hunk 1 above)
Comment (michaelspedersen) [reply bin/k8sinfra:157]: In the file it looks as follows - I think that's correct?
```
      else
        if @cluster_hash['k8s_infra']["release_type"]=="kubespray" then
	  latest = ks.latest_kubespray_release
	  ks.update_kubespray(latest)
        end
```

File: bin/k8sinfra (hunk 2 above)
Comment (michaelspedersen) [reply bin/k8sinfra:158]: That can be done. What about the addition of Multus and CNI plugins in the installation? Should that be part of default, or should it be kept as a separate e.g. "release" similar to how it is with dynamic Kubespray/K8s now?

File: bin/k8sinfra (hunk 3 above)
Comment (michaelspedersen) [reply bin/k8sinfra:247]: It is used to ensure the configuration file prints properly. I noticed lines were combined without it. I added it here as it is only a problem when `kubernetes_release` and `stable_k8s_release` are fetched dynamically.

**Comment by michaelspedersen** [comment]
Previous workarounds for Kubespray v2.12 have been removed, since they are all fixed with the most recent v2.13

**Pull request merged**